instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/EmbeddedInstrumentationVersions.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.internal;
+
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.list;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.logging.Level.FINE;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * This class is internal and is hence not for public use. Its APIs are unstable and can change at
+ * any time.
+ */
+public final class EmbeddedInstrumentationVersions {
+
+  private static final Map<String, String> versions = loadVersions();
+
+  private static Map<String, String> loadVersions() {
+    Map<String, String> versions = new HashMap<>();
+
+    // preload this once
+    try {
+      List<URL> urls =
+          list(
+              new ResourceLoaderProxy()
```

Comment:
Do I understand correctly that for library instrumentation this class needs to be in the same class loader as the library instrumentation so that it can find the versions file?

Reply:
Yep - which I think _should_ normally happen: library instrumentation triggers `Instrumenter` load, `Instrumenter` triggers this.

Reply:
What if you are using a library instrumentation along with the agent. Then this class will be in boot loader and won't find the library versions.

Reply:
The shaded version of this class (provided by the agent) will be in the bootloader; the unshaded version (provided by the application itself) should still see its own versions file (+ all other versions files, since they get resource-injected into all classloaders).

Reply:
Completely forgot that this would also get shaded. I still think it is possible to break it by placing instrumentation-api in parent class loader and library instrumentation in child class loader.

Reply:
Yes, that scenario would break this mechanism. I wonder how common it is though, you'd have to deliberately separate the instrumentation-api from the instrumentation.